### ISSUE TITLE
3.9+ PEP 585 - type hints for builtin types

### ISSUE THREAD

**Issue by terrdavis** [beartype/beartype#4]
In case you haven't seen this:
https://www.python.org/dev/peps/pep-0585/

Relevant:
```python
>>> isinstance([1, 2, 3], list[str])
TypeError: isinstance() arg 2 cannot be a parameterized generic
````

**Comment by leycec** [comment]
**Fantastic.** I *hadn't* seen that, as ill-gotten fate would have it. If a typing-related PEP isn't explicitly listed in the ["PEP 484 & Friends"](https://github.com/beartype/beartype#pep-484-friends) subsection of [README.rst](https://github.com/beartype/beartype/blob/master/README.rst), it's a safe bet that I'm shamefully clueless and desperately require your personal guidance.

So, thanks a bunch for giving it. I'll promptly add a reference to [PEP 585 -- Type Hinting Generics In Standard Collections](https://www.python.org/dev/peps/pep-0585) to that list of everything else we currently don't support.

I should also note that, unlike [PEP 563 -- Postponed Evaluation of Annotations](https://www.python.org/dev/peps/pep-0563) which I loathe with all my autoimmune-ridden being, I surprisingly like [PEP 585](https://www.python.org/dev/peps/pep-0585). Clearly, *this* is how they should have implemented generics from the get-go. Creating a parallel `typing` hierarchy for concrete types (e.g., `List` for `list`, `Tuple` for `tuple`) *never* made the slightest sense, given that those types already exist and should have just been reused as is – which PEP 585 now enables.

Sadly, since most existing type annotations leverage the now-obsolete `typing.List[str]` style rather than the now-sensible `list[str]` style, we'll (of course) have to support both. Cue [xkcd standards](https://xkcd.com/927).

**Comment by leycec** [comment]
:boom:

**Comment by leycec** [comment]
**Yikes.** Despite deprecating most of the existing public API of the `typing` module by 2025, [PEP 585 is officially unimplemented under *all* existing Python versions as of this writing](https://stackoverflow.com/questions/62871524/is-pep-585-unusable-at-runtime-under-python-3-7-and-3-8) – meaning Python 3.5—3.8.

That means it's currently unusable, despite advertising itself as usable under Python 3.7 and 3.8 via the standard `from __future__ import annotations` pragma. Someone sure dropped the ball on this one, folks.

I don't know why there's effectively no peer review, quality assurance, or stress testing for Python Enhancement Proposals (PEPs). Clearly, *something* is fundamentally awry with the existing process by which PEPs are submitted, reviewed, and accepted. Whatever's wrong, it's probably long been institutionalized and hard-baked into the cultural climate within the cadre of core Python developers. <sup>Good luck changing *that,* ever.</sup>

That isn't to say that we won't eventually support PEP 585. We absolutely will. But we'll need to wait for a usable, working implementation to drop before we can even begin toying around with ideas, implementations, and optimizations. Let's see what fallacious wonders Python 3.9 brings! :sparkler: 

Until then, **this remains open.**

**Comment by leycec** [comment]
Oh, and the long-awaited `beartype 0.2.0` stable release is officially out. We [now support most `typing` attributes of interest](https://github.com/beartype/beartype#features) except [the stupidly hard ones like `typing.Generic` and `typing.TypeVar`](https://github.com/beartype/beartype#no-compliance). Ain't nobody got spare volunteer time for that kinda non-triviality (at the moment, of course).

**Up next:** deep type-checking of `typing.List` and `typing.Sequence` containers in the upcoming `beartype 0.2.1`, which will probably drop sometime towards the end of September. *Let's do this!* :bear:

**Comment by albanie** [comment]
So excited for this.

**Comment by leycec** [comment]
28 days ago, I promised I would do a thing. Today, I complete that promise by doing a thing. Introducing...

**Beartype 0.3.0,** dangerously released into the radioactive [GitHub](https://github.com/beartype/beartype/releases/tag/v0.3.0) and [PyPI](https://pypi.org/project/beartype/0.3.0) badlands but moments ago. This release reveals the pure win of `O(1)` runtime type-checking through on onslaught of quasi-religious revelations, including...

## Deep Type-checking of Nested Sequences in O(1) Time: *Oh, It's Deep*

`@beartype` now type-checks both the shallow types *and* deep contents of parameters and return values annotated with `typing`-based sequences in guaranteed constant time. To do so, we exploit a variant of the [well-known coupon collector's problem](https://en.wikipedia.org/wiki/Coupon_collector%27s_problem) by randomly type-checking one item at each nesting level of each sequence on each call of each decorated callable. This then enables us to [statistically predict the number of calls required to fully type-check all items of an arbitrary container on average](https://github.com/beartype/beartype#id8). <sup>**Math spoiler alert:** *it's O(nlogn) calls.*</sup>

Currently supported `typing`-based sequences include:

* `typing.List`.
* `typing.MutableSequence`.
* `typing.Sequence`.

Since beartype 0.2.0 previously implemented deep type-checking for `typing.Optional` and `typing.Union`, we now deeply type-check all possible permutations of these type hints: e.g.,

```
from beartype import beartype
from typing import List, MutableSequence, Sequence, Union

@beartype
def hissing_nest_of_vipers(scary_vipers: Union[bool, str, Sequence[List[MutableSequence[float]]]) -> float:
     return my_floaty_tuple[0]

# These all totally work. Would beartype lie?
hissing_nest_of_vipers(True)
hissing_nest_of_vipers('They really taste great.')
hissing_nest_of_vipers(([[10.23, 9.24,], [8.25, 43.62,],], [[42.63,], [41.18,],]))

# This raises an exception. Beartype never lies! Never!
hissing_nest_of_vipers(([['Like, maybe don't put in your hand in there.'],],))
```

`typing.Tuple` and `typing.NamedTuple` are conspicuously absent with respect to deep type-checking, as their cray-cray syntax requires special-purposed handling that merits its own stable release (e.g., `typing.Tuple[wut, even, is, happening, here, bruh]`). *More on that later.*

Of course, tuples and named tuples are both immutable sequences. This means that **you can already deeply type-check tuples and named tuples in `beartype` 0.3.0** by simply annotating parameters and return values as `typing.Sequence` instead: e.g.,

```
from beartype import beartype
from typing import Sequence

@beartype
def accept_my_floaty_tuple(my_floaty_tuple: Sequence[float]) -> float:
     return my_floaty_tuple[0]

# This totally works -- even without suspicious squinting, though that helps.
accept_my_floaty_tuple((0.1, 3.6, 2.7, 13.20, 12.21, 11.22,))
```

### Optimized for His and Her Computational Pleasure under Python ≥ 3.8

Let's talk about speed, 'cause everyone likes it fast.

Under Python ≥ 3.8, `@beartype` generates optimal code when deeply type-checking items contained in nested subsequences (e.g., `List[List[List[str]]]`). Of course, we use the unclean dirty magic of [PEP 572-style assignment expressions](https://www.python.org/dev/peps/pep-0572) to do so. Yes, that's right, friends! PEP 572, the controversial syntactic change that prompted Guido to voluntarily abdicate his ancestral leadership role as BDFL, viciously lashes out again. Thanks to beartype 0.3.0, assignment expressions have finally proven their real-world utility. <sup>*Suck it, [python-ideas](https://mail.python.org/mailman3/lists/python-ideas.python.org)!*</sup>

I've profiled this eye-, soul-, and career-destroying optimization to yield speedups of **at least 200%** as the equivalent unoptimized code generated under Python < 3.8. For doubly- and triply-nested sequences (e.g., `List[List[str]]` and `List[List[List[str]`), it's about 200% – but the performance gap only increases as the nesting level increases. A doubling in performance is only the minimum improvement under Python ≥ 3.8. Sequences more than triply-nested (e.g., `List[List[List[List[List[str]]]]`) should expect a comparably dramatic speedup, though it *does* seem a tad doubtful that anyone in the history of our plucky species would ever commit that kind of anarchic `typing` monstrosity. Still, it's there. That's nice.

## Deep Human-readable Exception Messages: *They're Super-deep*

`@beartype` now raises human-readable exceptions exhibiting the exact cause(s) of deep type-checking failures, including failures deeply nested in subcontainers: e.g.,

> @beartyped pep_hinted() parameter pep_hinted_param=[([37],)] violates PEP type hint typing.List[typing.Sequence[typing.MutableSequence[str]]], as list item 0 tuple item 0 list item 0 value "37" not str.

Yeah. That's deep like my old soul. Currently supported `typing` type hints that now generate deep exception messages include:

* `typing.List`.
* `typing.MutableSequence`.
* `typing.Optional`.
* `typing.Sequence`.
* `typing.Union`.

A win for readable typing is a win for all mankind.

## The Dark Road Ahead

As America stands on the precipice of a renewed national unity and good tidings of brotherly love, so too do we chart the unmarked trail forwards to a brighter beartyped future:

### Beartype 0.4.0: It's Tuples All the Way Down

To maybe be released by late October, **Beartype 0.4.0** will *probably* implement support for deeply type-checking `typing.Tuple` and `typing.NamedTuple`, mostly because that's the minimal effort feature request that everyone (...which basically just means me and you, @albanie) wants.

A sample size of 2 is still a sample size. ...*what*?

### Beartype 0.5.0: NumPy Arrays, Eh?

To maybe be released by late November, **Beartype 0.5.0** will *probably* implement support for shallowly type-checking NumPy arrays via a newly introduced `beartype.cave.NumpyArrayType` thingy, because we desperately need this in both [BETSE](https://gitlab.com/betse/betse) and our data science and machine learning consultancy. Of course, NumPy arrays deeply type-check themselves by definition. So, there's no need for that here. What *is* there a need for then?

**NumPy array metadata,** including:

* **[Dtype](https://numpy.org/doc/stable/reference/arrays.dtypes.html)**. Some dtypes like [structured dtypes](https://numpy.org/doc/stable/user/basics.rec.html) and [the `object` dtype](https://stackoverflow.com/questions/29877508/what-does-dtype-object-mean-while-creating-a-numpy-array) dramatically alter usage patterns. So, it only makes sense for callables to want to constrain the dtypes of the arrays they accept as parameters and/or return.
* **[Shape](https://numpy.org/devdocs/reference/generated/numpy.shape.html).** While real-world usage patterns typically treat the size of each dimension of a NumPy array (i.e., the number of items in that dimension) as unconstrained free variables, the number of these dimensions is typically constrained at callable definition time. For example, a callable expecting a two-dimensional NumPy array *usually* doesn't particularly care how many rows or columns that array has – but it _really, really, **really*_ cares that that array has exactly two dimensions and no more or less.

So, I'm thinking preliminary syntax like this:

```
from beartype import beartype
from beartype.cave import NumpyArrayType

# This constrains the passed "funbag_of_fun" parameter to be a two-dimensional
# NumPy array with any float dtype (e.g., "np.float32", "np.float64").
@beartype
def catch_the_funbag(funbag_of_fun: NumpyArrayType[float, 2]) -> float:
     return funbag_of_fun[0, 0]
```

Of course, there currently exists no `typing` standard for annotating NumPy arrays – but since when did a lack of accepted standards, community debate, or common sense ever stop us? Never, of course! Those things, we need them not. Since NumPy is a third-party package (albeit a third-party package that's basically mandatory with all Python installs in 2020), it kinda seems doubtful there ever *will* exist a `typing` standard for annotating NumPy arrays.

That's where we come in. Like, us. You and me. It's kinda lonely in here, but a sample size of 2 is still a sample size. ...*what*?

### Beartype 0.6.0: *The Mappings Are Not the Territory*

To maybe be released by late December, **Beartype 0.6.0** will *probably* implement support for deeply type-checking `typing.Dict`, `typing.Mapping`, and `typing.MutableMapping`. Those are the last big-ticket items we need, but they're also the least trivial. Although the C-based CPython implementation almost certainly stores dictionary keys and values as hash bucket sequences, it doesn't expose those sequences to the Python layer. That means we have no efficient random access to arbitrary dictionary keys or values from within `@beartype`.

Does that complicate `O(1)` runtime type-checking of dictionaries? Yes. Yes, it does. Which is why we're saving dictionaries for last. I *do* have a number of risky ideas here, most of which revolve around internal caches of key and value iterators (i.e., the memory views returned by the `dict.keys()` and `dict.values()` methods). I don't want to blow anything up, so this requires care, forethought, and a rusty blood-flecked scalpel.

### Beartype 0.7.0: *Who Can Tell the Future, Controls the Future*

**Beartype 0.7.0** is currently open to creative interpretation. Whatever else [BETSE](https://gitlab.com/betse/betse) needs, beartype 0.7.0 gets.

Beyond that, the future is a smoking ruin of lost causes, dead dreams, and misspent middle-aged-hood. If you have any `typing` requests, <sup>*please not `TypeVar`, please not `TypeVar`</sup> speak now or forever hold your beartype feature list. :speaking_head:

**Comment by albanie** [comment]
This is great. 

I particularly like the idea for future NumPy validation. I face essentially the same use-case in PyTorch.  Even simple dimension checking to prevent (to catch broadcasting errors) could save buckets of tears for future generations.  

As an almost relevant aside, there have been some attempts to transition to "NamedTensors" (originally discussed here http://nlp.seas.harvard.edu/NamedTensor, and later introduced experimentally here https://pytorch.org/docs/stable/named_tensor.html). If they can iron out the implementation issues, in the long term these could lead to a significant improvement in readability and would be well-suited to type/shape-annotations.

**Comment by leycec** [comment]
> Even simple dimension checking to prevent (to catch broadcasting errors) could save buckets of tears for future generations.

Right? There's nothing worse than obscure non-human-readable exceptions arising like rancid burps from the bowels of higher-level APIs like SciPy or TensorFlow just because somewhere down the line some lower-level API accidentally generated a NumPy array with horrifying structure, which nobody bothered to check... *because they currently can't.*

I'm still not wedded on the syntax, but – after some calculated reflection and pensive soul-searching – I'm unconvinced that `typing`-style array subscription (e.g., `NumpyArrayType[float, 2]`) is the RightWay™ to declare array metadata. I mean, sure, it sorta works if you squint at it. But it's also non-human-readable. I can only envision endless discussions like:

> *2? For the love of Guido, what in the five hells of PEP 8 does 2 mean, Jim?*

> *...search me, Bob. Thought you knew. Must be a typo. I'll just remove that.*

It's also fragile and non-forward compatible. What if we want to add *yet another* type of array metadata whose value can also be an integer? Then... we're boned, in the bad way! That's bad.

I'm considering standard instantiation syntax with human-readable keywords instead ala:

```
from beartype import beartype
from beartype.cave import NumpyArrayType

# This constrains the passed "funbag_of_fun" parameter to be a two-dimensional
# NumPy array with any float dtype (e.g., "np.float32", "np.float64").
@beartype
def catch_the_funbag(funbag_of_fun: NumpyArrayType(dtype=float, ndim=2)) -> float:
     return funbag_of_fun[0, 0]
```

`NumpyArrayType(dtype=float, ndim=2)` instead of `NumpyArrayType[float, 2]` seems sane and sensible to me, but I also like [obscure Japanese visual novels](https://github.com/leycec/fsnrnue) and [obscure technical death metal albums from Quebec literally named **Obscura**](https://rateyourmusic.com/release/album/gorguts/obscura). My opinion may not be representative of mainstream consensus. :man_singer: 

> As an almost relevant aside, there have been some attempts to transition to "NamedTensors"...

You have my attention. I like what I read and read what I like, especially:

> In addition, named tensors use names to automatically check that APIs are being used correctly at runtime, providing extra safety.

The [*Name propagation semantics*](https://pytorch.org/docs/stable/named_tensor.html#name-propagation-semantics) section is especially exciting, suggesting it wouldn't be hard to *also* add support for PyTorch named tensors in a future `beartype` release – but probably *not* before `beartype` 0.7.0, unless someone who is not me would care to contribute a pull request. Any takers? Anyone? *Helloooooooo?*

BETSE currently only leverages NumPy arrays rather than PyTorch tensors – which, in hindsight, might have been a mistake. Who can argue with GPU acceleration? Let's put those shader units to good use, boys!

**Comment by leycec** [comment]
*Fire it up!* Beartype 0.3.2 featuring **Python 3.9 support** is now out. This is just a point compatibility release (with not much else to show for it), but at least we continue to uphold the golden future of Python.

Let us also express our gratitude to the Universe for Python 3.5's official death several weeks ago. Long live Python 3.6 f-strings, which we can now finally use. It's been a long road we've walked together, `str.format()`, but the bell tolls at long last for thee. :bell:

**Comment by leycec** [comment]
**Mondo hype generator:** the next stable release of beartype is but a few days away, stalwart `typing` heroes! **Beartype 0.4.0** celebrates a partial return to democratic sanity in Canada's southern neighbour by bringing:

* Deep type-checking of:
  * `typing.Tuple`, `typing.Annotated`, `typing.NewType`, and `typing.NoReturn`.
  * **Forward references** (i.e., unqualified relative classnames typically referring to user-defined classes that have yet to be defined).
  * **Generics** (i.e., classes subclassing one or more `typing` pseudo-superclasses), including both user-defined generics and standard `typing` generics such as:
    * `typing.IO`.
    * `typing.BinaryIO`.
    * `typing.TextIO`.
  * **Protocols** (i.e., classes directly subclassing the `typing.Protocol` abstract base class (ABC) *and* zero or more `typing` non-class objects), including both user-defined generics and standard `typing` protocols such as:
    * `typing.SupportsAbs`.
    * `typing.SupportsBytes`.
    * `typing.SupportsComplex`.
    * `typing.SupportsInt`.
    * `typing.SupportsFloat`.
    * `typing.SupportsRound`.
* Full compliance with:
  * [PEP 483 -- The Theory of Type Hints](https://www.python.org/dev/peps/pep-0483).
  * [PEP 484 -- Type Hints](https://www.python.org/dev/peps/pep-0484).
  * [PEP 544 -- Protocols: Structural subtyping (static duck typing)](https://www.python.org/dev/peps/pep-0544).
  * [PEP 560 -- Core support for typing module and generic types](https://www.python.org/dev/peps/pep-0560).
  * [PEP 563 -- Postponed Evaluation of Annotations](https://www.python.org/dev/peps/pep-0563).
  * [PEP 593 -- Flexible function and variable annotations](https://www.python.org/dev/peps/pep-0593).

With respect to [PEP 484](https://www.python.org/dev/peps/pep-0484), "full compliance" means "We deeply type-check a lot and at least shallowly type-check the remainder, yo." With respect to all of the *other* PEPs listed above, "full compliance" means "We deeply type-check everything, bro."

Notably, this means that beartype should no longer raise exceptions when decorating callables annotated by type hints compliant with *any* of the above PEPs. Beartype will still raise exceptions when decorating callables annotated by PEP 585-compliant type hints (e.g., `def uhoh(lawks: list[str]): pass`), but we hope to trivially resolve that as well by fully supporting PEP 585 with the next major beartype release: **beartype 0.5.0,** expected to be released at the end of this month.

We're on the cusp of full, *really* full compliance, folks. Once PEP 585 support lands, beartype will be both the most compliant and stupidly fastest Python runtime type checker in existence. The only substantial feature beartype lacks going forward is support for type-checking **type variables** (i.e., `typing.TypeVar` instances), which we currently silently ignore because we're not actually convinced anyone cares about type variables, they're fairly troublesome to support, and we don't actually require them ourselves. <sup>*Fight me.*</sup>

I'm painfully tired tonight and tomorrow's a family day. So, I hope to push beartype 0.4.0 out this Wednesday. The commits are in and all's passing. I mostly just need to sanity-check everything and cobble together the release notes. Thanks for everything, Pythonistas! Everyone here rocks it.

Disgustingly patriotic American icons for all: :us: :eagle: :us:

**Comment by leycec** [comment]
**"This is my boomstick."** Beartype 0.4.0 is now officially out on all the standard channels.

## tl;dr

This stable release significantly improves beartype's compliance with Python Enhancement Proposals (PEPs), including full compliance with:

  * [PEP 483 -- The Theory of Type Hints](https://www.python.org/dev/peps/pep-0483).
  * [PEP 484 -- Type Hints](https://www.python.org/dev/peps/pep-0484).
  * [PEP 544 -- Protocols: Structural subtyping (static duck typing)](https://www.python.org/dev/peps/pep-0544).
  * [PEP 560 -- Core support for typing module and generic types](https://www.python.org/dev/peps/pep-0560).
  * [PEP 563 -- Postponed Evaluation of Annotations](https://www.python.org/dev/peps/pep-0563).
  * [PEP 593 -- Flexible function and variable annotations](https://www.python.org/dev/peps/pep-0593).

With respect to PEPs 483 and 484, "full compliance" means "beartype deeply type-checks a lot and at least shallowly type-checks the remainder." With respect to the remaining PEPs, "full compliance" means "beartype deeply type-checks everything."

This sort of accidentally became a doctoral thesis-like monstrosity at some point. I'm unconvinced anyone actually wants to read this minutiae, but... for those with the patience of an Ecuadorian three-toed sloth, <sup>*...hi!*</sup> here's the slo-mo low-down.

## Compatibility Broken

* **Call-time type hint exception classes renamed.** For orthogonality with decoration-time type hint exception classes, all call-time type hint exception classes have been renamed from `beartype.roar.BeartypeCallCheck*Exception` to `beartype.roar.BeartypeCallHint*Exception` (e.g., from `beartype.roar.BeartypeCallCheckPepParamException` to `beartype.roar.BeartypeCallHintPepParamException`).

## Features Added

* **Deep `typing.Annotated` type-checking.** Under Python ≥ 3.9, the `@beartype` decorator now deeply type-checks parameters and return values annotated by PEP 593 (i.e., "Flexible function and variable annotations")-compliant `typing.Annotated` type hints in guaranteed constant time.
* **Deep `typing.Generic` type-checking.** The `@beartype` decorator now type-checks both the shallow types *and* deep contents of parameters and return values annotated by PEP 484-compliant **generics** (i.e., type hints subclassing a combination of one or more of the `typing.Generic` superclass and/or other `typing` non-class pseudo-superclasses) in guaranteed constant time. Specifically, beartype iteratively walks up the superclass hierarchy of each generic and deeply type-checks that the current parameter or return value satisfies all constraints implied by that superclass. This includes:
  * `typing.IO`. This was no small feat, as this abstract base class (ABC) fails to leverage structural subtyping (e.g., via the PEP 544-compliant `typing.Protocol` ABC) and is thus unusable at runtime... like most `typing` objects, sadly. Our solution is to ignore the existing implementation of these classes, declare our own internal `typing.Protocol`-based variants of these classes, and implicitly substitute all instances of these classes with our own variants during our breadth-first traversal (BFS) over PEP-compliant type hints.
  * `typing.BinaryIO`.
  * `typing.TextIO`.
* **Deep `typing.NewType` support.** The `@beartype` decorator now deeply type-checks parameters and return values annotated by PEP 484-compliant **new types** (i.e., closures generated by the `typing.NewType` closure factory) in guaranteed constant time.
* **Deep `typing.NoReturn` support.** The `@beartype` decorator now fully type-checks that callables annotated by the `typing.NoReturn` singleton return *no* values (i.e., either halt the active Python process or raise an exception) in guaranteed constant time.
* **Deep `typing.Protocol` type-checking.** Under Python ≥ 3.8, the `@beartype` decorator now type-checks both the shallow types *and* deep contents of parameters and return values annotated by PEP 544 (i.e., "Protocols: Structural subtyping (static duck typing)")-compliant **protocols** (i.e., type hints subclassing a combination of one or more of the `typing.Protocol` superclass and/or other `typing` non-class pseudo-superclasses) in guaranteed constant time, similarly to how beartype type-checks generics. This includes:
  * `typing.SupportsAbs`.
  * `typing.SupportsBytes`.
  * `typing.SupportsComplex`.
  * `typing.SupportsIndex`.
  * `typing.SupportsInt`.
  * `typing.SupportsFloat`.
  * `typing.SupportsRound`.
* **Deep `typing.Tuple` type-checking.** The `@beartype` decorator now type-checks both the shallow types *and* deep contents of parameters and return values annotated by PEP 484-compliant `typing.Tuple` type hints in guaranteed constant time. The three syntactic variants standardized by PEP 484 are all supported, including:
  * `typing.Tuple[()]`, type-checking empty tuples.
  * `typing.Tuple[{typename}, ...]`, type-checking tuples containing arbitrarily many items all satisfying the PEP-compliant child hint `{typename}` (e.g., `typing.Tuple[str, ...]`, a tuple of strings).
  * `typing.Tuple[{typename1}, ..., {typenameN}]`, type-checking tuples containing exactly `N` items each satisfying a unique PEP-compliant child hint `{typenameI}` (e.g., `typing.Tuple[str, int, float]`, a tuple containing exactly one string, one integer, and one floating-point number in that order).
* **Shallow `typing.AsyncContextManager`, `typing.ContextManager`, `typing.Match`, and `typing.Pattern` support.** The `@beartype` decorator now type-checks the shallow types (but *not* deep contents) of parameters and return values annotated by the PEP 484-compliant `typing.AsyncContextManager`, `typing.ContextManager`, `typing.Match`, and `typing.Pattern` objects. This was no small feat, as the implementations of:
  * `typing.AsyncContextManager` and `typing.ContextManager` define ``__repr__()`` dunder methods returning erroneous machine-readable representations obstructing proper support under prior releases.
  * `typing.Match` and `typing.Pattern` under Python 3.6 are non-trivially obtuse and basically broken at runtime.
* **Shallow type variable support.** The `@beartype` decorator now shallowly type-checks all otherwise supported PEP-compliant type hints parametrized by one or more type variables (e.g., `List[T]`, where `T = TypeVar('T')`). Previously, this decorator raised exceptions when decorating callables with such hints. This decorator does *not* yet deeply type-check type variables to be constrained across callable parameters and return values or class methods. For `def muh_func(muh_param: List[T]) -> T: return muh_param[0]`, as example, this decorator now shallowly type-checks the passed parameter `muh_param` to be a list but does *not* yet deeply type-check that this function returns values of the same types as items of this list.

## Features Optimized

* **Nested PEP-compliant type hint type-checking.** Under Python ≥ 3.8, the `@beartype` decorator now generates optimized code when deeply type-checking items contained in arbitrarily nested PEP-compliant type hints (e.g., `typing.List[typing.Union[bool, str, typing.List[int]]]`) with PEP 572-style assignment expressions, generalizing similar sequence-specific optimizations introduced with beartype 0.3.0.
* **PEP-compliant type hint breadth-first search (BFS).** The `@beartype` decorator now internally catalogues PEP-compliant type hints during the breadth-first search (BFS) it performs over these hints as simple tuples rather than fixed lists, as the former are both mildly faster than the latter and significantly more maintainable (which is the main gain here).

## Issues Resolved

* **PEP-compliant type hint code generation memoization.** The `@beartype` decorator now properly memoizes the type-checking code it generates for PEP-compliant type hints. Prior beartype releases silently failed to memoize this code across different callables annotated by the same hints.

## API Changed

* In the `beartype.cave` submodule, added:
  * **A new public `beartype.cave.HintPep585Type` attribute**, defined as the C-based type of all PEP 585-compliant type hints (i.e., C-based type hint instantiated by subscripting either a concrete builtin container class like `list` or `tuple` *or* an abstract base class (ABC) declared by the `collections.abc` submodule like `collections.abc.Iterable` or `collections.abc.Sequence`) if the active Python interpreter targets at least Python 3.9.0 *or* `beartype.cave.UnavailableType` otherwise. This is a prerequisite for PEP 585 support in a subsequent stable release.
* In the `beartype.roar` submodule, renamed:
  * `BeartypeCallCheckException` to `BeartypeCallHintException`.
  * `BeartypeCallCheckPepException` to `BeartypeCallHintPepException`.
  * `BeartypeCallCheckPepParamException` to
    `BeartypeCallHintPepParamException`.
  * `BeartypeCallCheckPepReturnException` to
    `BeartypeCallHintPepReturnException`.
  * `BeartypeCallCheckNonPepException` to
    `BeartypeCallHintNonPepException`.
  * `BeartypeCallCheckNonPepParamException` to
    `BeartypeCallHintNonPepParamException`.
  * `BeartypeCallCheckNonPepReturnException` to
    `BeartypeCallHintNonPepReturnException`.

*Phew.* :sweat_smile:

**Comment by albanie** [comment]
Glorious. Lots of useful stuff in here for me 😄 .

At the risk of potentially adding further workload, perhaps at some point in future (no rush, obviously), you could add a short technical report so that the `beartype` enthusiasts among us can cite/credit the library.

**Comment by leycec** [comment]
That... *would* be useful. My wife and I currently lack University affiliation, which complicates publication just a tad. We've both walked the independent scholar route before, but that's neither a fun nor easy journey. Without affiliation, getting even simple letters to the editor or literature reviews past journal editors out for peer review is (perhaps justifiably) an uphill journey of gatekeeping and rejection and dejection and really non-happy feelings.

That said, you're right – as always, eh? `beartype` has a few novel theoretical aspects that would warrant publication. It'd be especially nice to see if there isn't a tighter bound than [`O(n log n)` for the expected number of calls required to deeply type-check an arbitrary container of `n` items](https://github.com/beartype/beartype#nobody-believes-you), though I sorta doubt I still have the statistical R chops to get that properly done. `</sigh>`

We're both contemplating a change of occupational scenery next year. A furious global pandemic isn't exactly the optimal job market, but what can you do? With any luck (and more than a little hand-wringing), one or both of us will reignite our collaborative passion for theoretical research and hook back up with that whole academia scene!

**Comment by leycec** [comment]
**Boom-shaka!** `beartype 0.5.0` just mic-dropped. Full support for [PEP 585](https://www.python.org/dev/peps/pep-0585) along with a host of ancillary compatibility improvements and optimizations is in. Since this *finally* resolves this long-standing issue, this is all she wrote for issue #4, folks. Let's do this!

# Changelog for Great Glory

**Beartype 0.5.0** released.

This stable release implements full compliance for [PEP 585 -- Type Hinting Generics in Standard Collections](https://www.python.org/dev/peps/pep-0585), where "full compliance" means "`@beartype` deeply type-checks all categories of type hints deeply type-checked by prior stable releases and shallowly type-checks the remainder." See our [compliance list](https://github.com/beartype/beartype#full-compliance) and [feature matrix](https://github.com/beartype/beartype#features) for deeply exhausting, enervating, and deenergizing details.

Specific changes include:

## Compatibility Improved

* **[PEP 585](https://www.python.org/dev/peps/pep-0585).** `@beartype` is now fully compliant with [PEP 585 -- Type Hinting Generics in Standard Collections](https://www.python.org/dev/peps/pep-0585). Since that PEP supercedes and largely obsoletes the overwhelming majority of [PEP 484](https://www.python.org/dev/peps/pep-0484), compliance with [PEP 585](https://www.python.org/dev/peps/pep-0585) is critical for *all* static and runtime type checkers to guarantee forward compatibility with Python's type-checking ecosystem. Notably, `@beartype` now provides deep type-checking support for all PEP 585-compliant type hints that had comparable deep type-checking support under prior `@beartype` releases, including:
    * Subscriptions of these C-based builtin types and pure-Python abstract base classes (ABCs) declared by the stdlib:
      * `list`.
      * `tuple`.
      * `collections.abc.ByteString`.
      * `collections.abc.MutableSequence`.
      * `collections.abc.Sequence`.
    * User-defined PEP 585-compliant generics (e.g., `class ListOfInts(list[int]): pass`), which were implemented in a completely non-orthogonal manner to both PEP 484-compliant generics and PEP 544-compliant protocols and thus required "extreme" handling for PEP 585-specific edge cases.
* **Unhashable PEP-compliant type hints.** `@beartype` now supports PEP-compliant type hints *not* hashable by the `hash()` builtin and thus impermissible for use as dictionary keys, set members, and memoized callable parameters. While *all* PEP 484-compliant type hints are hashable, many such hints  (e.g., `typing.Callable[[], str]`) are unhashable when converted into equivalent PEP 585-compliant type hints (e.g., `collections.abc.Callable[[], str]`), necessitating that `@beartype` now permissively accept both hashable and unhashable type hints. The disadvantage of the latter is that the private `@beartype._util.cache.utilcachecall.callable_cached` decorator underlying internal decoration-time memoization performed by `@beartype` cannot, by definition, memoize calls passed one or more unhashable objects. Ergo, callables accepting one or more unhashable PEP-compliant type hints incur a performance penalty versus hashable PEP-compliant type hints. Note this penalty is only paid at decoration rather than call time and *should* thus be entirely negligible. Unhashable PEP-compliant type hints include:
  * PEP 585-compliant type hints subscripted by one or more unhashable objects (e.g., `collections.abc.Callable[[], str]).
  * PEP 586-compliant type hints subscripted by an unhashable object (e.g., `typing.Literal[[]]`, a literal empty list).
  * PEP 593-compliant type hints subscripted by one or more unhashable objects (e.g., `typing.Annotated[typing.Any, []]`, the `typing.Any` singleton annotated by an empty list).
* `@typing.no_type_check`. `@beartype` now supports the PEP 484-compliant `@typing.no_type_check` decorator by silently ignoring (and thus reducing to a noop for) all callables decorated by that decorator.
* `typing.TYPE_CHECKING.` `@beartype` now supports the PEP 484-compliant `typing.TYPE_CHECKING` boolean constant by silently reducing to the identity decorator when this boolean is ``True`` (i.e., during external static type checking).

## Compatibility Broken

* **Call-time PEP-noncompliant type hint exception classes removed.** Since *all* PEP-noncompliant type hints supported by `@beartype` (e.g., tuple unions) are now internally coerced into equivalent PEP-compliant type hints, PEP-noncompliant type hint exception classes are now obsolete and have been summarily removed. These include:
  * `beartype.roar.BeartypeCallHintNonPepException`.
  * `beartype.roar.BeartypeCallHintNonPepParamException`.
  * `beartype.roar.BeartypeCallHintNonPepReturnException`.

## Features Deprecated

* **PEP 484-compliant type hints deprecated by PEP 585.** `@beartype` now emits non-fatal warnings of class `beartype.roar.BeartypeDecorHintPepDeprecatedWarning` under Python >= 3.9 for each PEP 484-compliant type hint deprecated by PEP 585 annotating each decorated callables. Critically, note that this deprecates *most* PEP 484-compliant type hints accepted without warning by prior stable releases and, indeed, most of the existing `typing` stdlib module. Affected type hints include subscriptions of *any* of the following objects:
  * `typing.AbstractSet`.
  * `typing.AsyncGenerator`.
  * `typing.AsyncIterable`.
  * `typing.AsyncIterator`.
  * `typing.Awaitable`.
  * `typing.ByteString`.
  * `typing.Callable`.
  * `typing.ChainMap`.
  * `typing.Collection`.
  * `typing.Container`.
  * `typing.ContextManager`.
  * `typing.Coroutine`.
  * `typing.Counter`.
  * `typing.DefaultDict`.
  * `typing.Deque`.
  * `typing.Dict`.
  * `typing.FrozenSet`.
  * `typing.Generator`.
  * `typing.ItemsView`.
  * `typing.Iterable`.
  * `typing.Iterator`.
  * `typing.KeysView`.
  * `typing.List`.
  * `typing.MappingView`.
  * `typing.Mapping`.
  * `typing.Match`.
  * `typing.MutableMapping`.
  * `typing.MutableSequence`.
  * `typing.MutableSet`.
  * `typing.Pattern`.
  * `typing.Reversible`.
  * `typing.Sequence`.
  * `typing.Set`.
  * `typing.Tuple`.
  * `typing.Type`.
  * `typing.ValuesView`.

## Features Optimized

* **Standard PEP-noncompliant class type hints.** `@beartype` now efficiently registers type hints that are standard classes *not* explicitly compliant with any existing PEP at decoration time and accesses these hints at call time via the standard beartypistry singleton used to resolve all other type hints rather than inefficiently accessing these hints at call time via dictionary lookup into the ``__annotations__`` dunder attribute of the decorated callable, eliminating one dictionary lookup for each such hint for each call to each callable annotated by one or more such hints.
* **Beartype-specific tuple unions < PEP 484 unions.** Beartype-specific tuple unions are now a strict semantic subset of PEP 484 unions. Previously, tuple unions and PEP 484 unions were implemented in disparate subpackages -- and the subpackage implementing tuple unions has become poorly documented, maintained, and tested. To reduce bit-rot across the codebase and unify the implementation of tuple and PEP 484 unions, `@beartype` now silently coerces all tuple unions into the equivalent PEP 484 unions at decoration time. Doing so resolves all outstanding performance and usability issues with tuple unions, including:
  * Tuple unions are now efficiently registered at decoration time and accessed at call time via the standard beartypistry singleton used to resolve all other type hints rather than inefficiently accessed at call time via dictionary lookup on the ``__annotations__`` dunder attribute of the decorated callable, eliminating one dictionary lookup for each such hint for each call to each callable annotated by one or more such hints.
  * Duplicate types in tuple unions are now silently ignored rather than type-checked (e.g., ``(bool, str, bool)`` is now type-checked as ``(bool, str)``).
  * Forward references in tuple unions (e.g., ``('muh.TypeName', str)``) are now efficiently resolved at call time via the standard beartypistry singleton used to resolve all other forward references rather than inefficiently resolved at call time via non-standard iteration hard-coded into all wrapper functions generated by `@beartype` for callables annotated by these tuple unions.
  * Registration and optimized access of types in tuple unions via our internal beartypistry type registrar.
* **Decoration-time memoization,** particularly with respect to PEP 484 union type hints. All internal callables decorated by the private `@beartype._util.cache.utilcachecall.callable_cached` decorator are now efficiently called with positional arguments rather than inefficiently called with equivalent keyword arguments. Memoizing keyword arguments is substantially more space- and time-intensive than memoizing the equivalent positional arguments, partially defeating the purpose of memoization in the first place. To enforce this, these internal callables now emit new non-fatal private `beartype.roar._BeartypeUtilCallableCachedKwargsWarning` warnings when passed one or more keyword arguments.

## Issues Resolved

* **["3.9+ PEP 585 - type hints for builtin types."](https://github.com/beartype/beartype/issues/4)** This release resolves longstanding issue #4 by adding full support for PEP 585. Thanks for prodding us to jump onto this critical feature, @terrdavis!

## Tests Improved

* **Third-party package tests restored.** Unit tests conditionally depending upon the importability of third-party packages have been restored to their prior working state after an embarassingly elongated period of being ignored. These include:
  * The `test_api_cave_lib_numpy()` unit test exercising NumPy integration in the `beartype.cave` submodule.
  * The `test_api_cave_lib_setuptools()` unit test exercising `setuptools` integration in the `beartype.cave` submodule.

## API Changed

* In the `beartype.roar` submodule:
  * Added:
    * `BeartypeDecorHintPepDeprecatedWarning`.
  * Removed:
    * `BeartypeCallHintNonPepException`.
    * `BeartypeCallHintNonPepParamException`.
    * `BeartypeCallHintNonPepReturnException`.

**Comment by leycec** [comment]
**Peace out.** Have a hopefully COVID-free Holidays, everyone! :christmas_tree: :santa: :snowman_with_snow:

**Comment by amogorkon** [comment]
In https://github.com/amogorkon/justuse we came up with a different solution that should be future-proof and doesn't require renaming all those names everywhere if you ever wanted to switch to "real generics".
If you put 
```
from gc import getreferents
from typing import _GenericAlias as GenericAlias
for t in (list, dict, set, tuple, frozenset):
    r = get_referents(t.__dict__)[0]
    r.update({
        "__class_getitem__": classmethod(GenericAlias),
    })
```
in your `__init__.py` or some other location that's called very early on, it should patch the built-in generic types from python versions down to 3.7 where `__class_getitem__` was introduced formally (but for some reason never added to the generics..), thus basically "backporting" the feature.

Since this approach makes use of the gc, it's CPython-specific, so you might have to tweak it a little if you want to make it work for other interpreters.

A fair warning: When we added this to our code, it interacted unpredictably and hard to debug with some other non-standard code that was lurking in some dark corner. So, at least expect the unexpected when trying this approach, but nothing bad happens, you won't have fear or even care about the deprecation of `typing` anymore.

**Comment by leycec** [comment]
Give @amogorkon a lever and he will move the Python world. Never would I even consider dangerously monkey-patching a `GenericAlias`-based `__class_getitem__()` implementation into Python's core C types. You've just shown me a darkness beyond all prior conception of darkness – *and I love it.*

Does that cauldron of insanity generalize to [everything else deprecated by PEP 585 too](https://www.python.org/dev/peps/pep-0585/#toc-entry-5), like:

* Abstract base classes published by `collections.abc` and `contextlib`?
* Concrete classes published by `collections` and `re`?

If so, I'm all in and will gleefully add this ~~eldritch horror~~ "clean resolution" to [our public knowledge base on circumventing PEP 585 deprecations](https://github.com/beartype/beartype#are-we-on-the-worst-timeline).

Unrelatedly, [`justuse`](https://github.com/amogorkon/justuse) looks *amazing.* Between one-off imports (e.g., `use("pprint").pprint(some_dict)` ← <sub>*wut*</sub>) and auto-installing imports (e.g., `np = use("numpy", version="1.21.0rc2", hashes={"3c90b0bb77615bda5e007cfa4c53eb6097ecc82e247726e0eb138fcda769b45d"}, modes=use.auto_install)` ← <sub>*wtf is even happening there*</sub>), I'm both disturbed and impressed – and that's hard to accomplish these days. Like most wilderness cabin-dwellers, I thrive on a daily diet of being jaded, bitter, and oppressively cynical.

Have you considered posting `justuse` to [/r/Python](https://reddit.com/r/Python)? Actually, I'm considering doing that for you. All your sweet Reddit karma will be mine. *Oh, yes.* :+1:

**Comment by amogorkon** [comment]
Since I'm not very active on Reddit and do *love* beartype, I will gladly relinquish the Reddit karma to you ;-)

Regarding your questions, I'm not entirely sure since we haven't actually tried to generalize it. Given that `__class_getitem__` is a general protocol, I wouldn't be surprised if you could probably put it *everywhere* and things would "just work".. or maybe it'll open a portal to another plane and release unspeakable horrors - we will only know for sure when we try it! For Science!

**Comment by TeamSpen210** [comment]
The classes in `collections.abc` and `contextlib` are just pure-Python classes, so you can set attributes normally. `__class_getitem__` is just an attribute lookup triggered when a regular `__getitem__` lookup on the metaclass fails, so it should just work fine if you can get an object into the dict.
